elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -239,9 +260,171 @@ export class SparseMerkleTree {
 		return bottomNode;
 	}
 
-	/*
-		public generateSingleProof() {}
-		public generateMultiProof() {}
-		public verifyMultiProof() {}
-		*/
+	public generateSingleProof = async (queryKey: Buffer): Promise<SingleProof> => {
+		const rootNode = await this.getNode(this._rootHash);
+		let currentNode = rootNode;
+		if (currentNode instanceof Empty) {
+			return {
+				key: queryKey,
+				value: EMPTY_VALUE,
+				binaryBitmap: bufferToBinaryString(EMPTY_HASH),
+				siblingHashes: [],
+				ancestorHashes: [],
+			};
+		}
+
+		let h = 0;
+		const siblingHashes = [];
+		const ancestorHashes = [];
+		let binaryBitmap = '';
+		const binaryKey = binaryExpansion(queryKey, this.keyLength);
+
+		while (currentNode instanceof Branch) {
+			ancestorHashes.push(currentNode.hash);
+			const d = binaryKey.charAt(h);
+			let currentNodeSibling: TreeNode = new Empty();
+			if (d === '0') {
+				currentNodeSibling = await this.getNode(currentNode.rightHash);
+				currentNode = await this.getNode(currentNode.leftHash);
+			} else if (d === '1') {
+				currentNodeSibling = await this.getNode(currentNode.leftHash);
+				currentNode = await this.getNode(currentNode.rightHash);
+			}
+
+			if (currentNodeSibling instanceof Empty) {
+				binaryBitmap = `0${binaryBitmap}`;
+			} else {
+				binaryBitmap = `1${binaryBitmap}`;
+				siblingHashes.push(currentNodeSibling.hash);
+			}
+			h += 1;
+		}
+
+		if (currentNode instanceof Empty) {
+			// exclusion proof
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: queryKey,
+				value: EMPTY_VALUE,
+			};
+		}
+
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') !== queryKey.toString('hex')
+		) {
+			// exclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') === queryKey.toString('hex')
+		) {
+			// inclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		return {
```

Comment:
in which branch in the LIP this happens?

Reply:
It's not supposed to happen, this condition should never reach. I added it because there's no default return statement for the function. Maybe there's another way around to default return ?

Reply:
i think logically `currentNode instanceof Leaf` is not needed because only possible case is the leaf.
Therefore, you could just check current `currentNode.key.equals(queryKey)` then return x, otherwise return y without last return